ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveInsertOnlyScanWriteIdRule.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite.rules.views;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.hadoop.hive.ql.io.AcidUtils;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.apache.hadoop.hive.ql.optimizer.calcite.RelOptHiveTable;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveTableScan;
+
+import static org.apache.hadoop.hive.conf.Constants.INSERT_ONLY_FETCH_BUCKET_ID;
+
+/**
+ * This rule turns on populating writeId of insert only table scans.
+ */
+public class HiveInsertOnlyScanWriteIdRule extends RelOptRule {
+
+  public static final HiveInsertOnlyScanWriteIdRule INSTANCE = new HiveInsertOnlyScanWriteIdRule();
+
+  private HiveInsertOnlyScanWriteIdRule() {
+    super(operand(HiveTableScan.class, none()));
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    HiveTableScan tableScan = call.rel(0);
+    Table tableMD = ((RelOptHiveTable) tableScan.getTable()).getHiveTableMD();
+    return !tableMD.isMaterializedView() && AcidUtils.isInsertOnlyTable(tableMD);
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    HiveTableScan tableScan = call.rel(0);
+    RelNode newTableScan = call.builder()
+            .push(tableScan.setTableScanTrait(HiveTableScan.HiveTableScanTrait.FetchInsertOnlyBucketIds))
```

Comment:
I suppose we have the information that a table is insert only before constructing the initial logical plan. Why do we need a rule to set this and we don't do it when constructing the initial `HiveTableScan`?

Reply:
WriteId is populated into the `ROW_ID` struct which contains three fields: `writeId, bucketId, rowId`. In case of insert-only tables `rowId` is not available so `FetchInsertOnlyBucketIds` is an internal feature.
The initial logical plan for MV rebuild is a full rebuild and it does not required writeIds.
Later when all check passed which is need for incremental rebuild the plan is transformed to an incremental rebuild plan and that also required `writeId` filtering.

Reply:
I assume that at some point (later in the compilation phase),  you check the `HiveTableScan` operator to obtain/use the `FetchInsertOnlyBucketIds` trait. Unless I am missing something you have all the information already inside the `HiveTableScan` (`!tableMD.isMaterializedView() && AcidUtils.isInsertOnlyTable(tableMD)`) to derive it on the fly when needed. I don't understand very well why you need a rule to set it explicitly. 

Apologies if my reasoning is off but this is my happy break from escalations :)

Reply:
I don't want to enable fetching writeId automatically for any non MV insert only table.
If someone execute the query like 
```
SELECT t1.ROW__ID, t1.ROW__ID.writeId, a, b FROM t1;
```
where `t1` is an insert only table 
the result will contain `0` for ROW__ID.rowId for all records but will contain valid values for bucketId and writeId but only for not compacted records. This is confusing for users.

But this feature can still be used in case of incremental MV rebuild since compaction excludes the use of incremental rebuild.

So the goal was enable writeId fetch just when it is necessary. To control it in the AST level the `insertonly.fetch.bucketid` is used. It is a similar property like the `acid.fetch.deleted.rows`.
See #2549

Reply:
OK, I finally got it :) Thanks for explaining.

Reply:
Please add a small comment in the Class javadoc cause I am sure my future self seeing this will not remember the discussion and may try to remove the rule or put it in another place :).

Reply:
Sure. Added these details to the javadocs.